examples/gps_bot/gps_bot.cc

```diff
@@ -0,0 +1,90 @@
+// BoB robotics includes
+#include "common/logging.h"
+#include "common/timer.h"
+#include "common/gps.h"
+#include "hid/joystick.h"
+#include "robots/norbot.h"
+#include "third_party/units.h"
+
+// Standard C++ includes
+#include <iostream>
+#include <thread>
+
+// Standard C includes
+#include <cmath>
+
+using namespace BoBRobotics;
+using namespace units::angle;
+using namespace units::length;
+
+void gpsThreadFunc(const char *device, std::mutex &mutex, degree_t &latOut, degree_t &lonOut, meter_t &altitudeOut)
+{
+    Gps gps("/dev/ttyACM0");
+
+    std::cout << "GPS created" << std::endl;
+    while(true)
+    {
+        
+        // get gps data
+        degree_t lat,lon;
+        meter_t altitude;
+        bool didGetGps = gps.getPosition(lat,lon, altitude);
+
+        {
+            std::lock_guard<std::mutex> g(mutex);
+            latOut = lat;
+            lonOut = lon;
+            altitudeOut = altitude;
+        }
+    }
+}
+int main()
+{
+    constexpr float joystickDeadzone = 0.25f;
+    
+    // Create joystick interface
+    BoBRobotics::HID::Joystick joystick(joystickDeadzone);
+    
+    // Create motor interface
+    BoBRobotics::Robots::Norbot robot;
+
+    // output file for gps 
+    std::ofstream file("out.csv");
+
+    Timer<> timer("");
+
+    degree_t lat,lon;
+    meter_t altitude;
+    std::mutex gpsMutex;
+    std::thread gpsThread(gpsThreadFunc, "/dev/ttyACM0", std::ref(gpsMutex), std::ref(lat), std::ref(lon), std::ref(altitude));
```

Comment:
You don't need the ``std::ref``s here; the compiler knows they have to be references based on the function declaration. It won't hurt though.

Reply:
Is that true? https://stackoverflow.com/questions/36341724/c-thread-taking-reference-argument-failed-compile and the example code at http://www.cplusplus.com/reference/thread/thread/thread/ suggest that it isn't true...

Reply:
Oops, my bad! Ignore me.